### ISSUE TITLE
Pygame window not closing in jupyter notebook (window hangs)

### ISSUE THREAD

**Issue by pseudo-rnd-thoughts** [pygame-community/pygame-ce#2256]
**Environment:**

- **Operating system** (e.g. Windows, Linux(Debian), Linux(Ubuntu), Mac): Mac
- **Python version** (e.g. 3.7.9, 3.8.5) : 3.9
- **SDL version** (e.g. SDL 2.0.12): 2.26.4
- **PyGame version** (e.g. 2.0.0.dev10, 1.9.6): 2.4.0 (and 2.1.3)
- **Relevant hardware** (e.g. if reporting a bug about a controller, tell us the brand & name of it):

**Current behavior:**

I'm part of the development team of [OpenAI Gym](https://github.com/openai/gym/) and [Gymnasium](https://github.com/farama-Foundation/gymnasium) and would like users to be able to use pygame in jupyter notebook.
We are able to open a window in jupyter notebook however `pygame.quit` doesn't cause the window to close and the kernel will hang and close

**Expected behavior:**

To be able to close a pygame windows in jupyter notebook

**Steps to reproduce:**

`pip install gymnasium[classic_control]`

```python
import gymnasium as gym

env = gym.make("LunarLander-v2", render_mode="human")
observation, info = env.reset()

for _ in range(100):
   action = env.action_space.sample()
   observation, reward, terminated, truncated, info = env.step(action)

   if terminated or truncated:
      observation, info = env.reset()
env.close()  # This includes `pygame.quit()` 
```

I have tested this is issue with pygame using the following basic script 
```python
import pygame

win = pygame.display.set_mode((500,500))
pygame.display.set_caption("First Game")

run = True
while run:
    pygame.time.delay(100)

    # handle events
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            run = False

    # clear the disaply
    win.fill(0)

    # draw the scene
    filled_rect = pygame.Rect(100, 100, 25, 25)
    pygame.draw.rect(win, (0,0,255), filled_rect)

    # update the dispaly
    pygame.display.update()

pygame.quit()
```

**Comment by Starbuck5** [comment]
Hello, thanks for the issue.

I was unable to replicate this on my Windows computer, I can dig up a Mac and see about that sometime.

**Comment by pseudo-rnd-thoughts** [comment]
From reports with users, I suspect that this is a mac only issue

**Comment by rschiewer** [comment]
Is there any update on this yet? Facing the same problem using a gymnasium environment rendered via pygame on MacOS.

**Comment by traversaro** [comment]
I am probably missing something, but it seems that gymnasium still uses `pygame` in https://github.com/Farama-Foundation/Gymnasium/blob/1e55aa7b900a3e73e7c81c1f81ad6fab7ddbf3b0/pyproject.toml#L37 and not `pygame-ce` ?

**Comment by pseudo-rnd-thoughts** [comment]
@traversaro The problem to my knowledge is to both pygame and pygame-ce however as ce is updated more regularly than pygame I hope I would see a solution here

**Comment by traversaro** [comment]
> @traversaro The problem to my knowledge is to both pygame and pygame-ce however as ce is updated more regularly than pygame I hope I would see a solution here

The part I miss: did anyone ever tried with pygame-ce instead of pygame? I guess gymnasium users report will only have installed the dependency of gymnasium, that is pygame, not pygame-ce .

**Comment by pseudo-rnd-thoughts** [comment]
@traversaro I've just tested the code above on macos arm, python 3.12 and pygame-ce 2.5.5 with the same problem occuring

**Comment by pseudo-rnd-thoughts** [comment]
@traversaro I got bored today so got fable to identify the root cause https://github.com/libsdl-org/SDL/pull/15992 

Here is a user-level fix
```python
import time

# instead of pygame.quit():
pygame.Window.from_display_module().destroy()   # window ordered out; animation starts
t0 = time.time()
while time.time() - t0 < 0.5:                   # service the run loop past the ~250ms animation
    pygame.event.pump()
    pygame.time.delay(20)
pygame.quit()
```

Or if PyGame wants a fix without SDL changing
```c++
  #ifdef __APPLE__                                                                                                                                                                                                                                                                                                          
      /* macOS animates window close (~250ms) and only removes the window                                                                                                                                                                                                                                                   
         from screen while the main runloop is serviced; pump past it so                                                                                                                                                                                                                                                    
         quit() doesn't leave a ghost window (issue #2256). */                                                                                                                                                                                                                                                              
      {                                                                                                                                                                                                                                                                                                                     
          Uint32 start = SDL_GetTicks();                                                                                                                                                                                                                                                                                    
          while (SDL_GetTicks() - start < 400) {                                                                                                                                                                                                                                                                            
              SDL_PumpEvents();                                                                                                                                                                                                                                                                                             
              SDL_Delay(20);                                                                                                                                                                                                                                                                                                
          }                                                                                                                                                                                                                                                                                                                 
      }                                                                                                                                                                                                                                                                                                                     
  #endif      
```

Or this much more cursed code that edits the SDL
```python
"""Workaround for pygame windows staying on screen after pygame.quit() on
macOS in environments that keep running afterwards (Jupyter, REPL).

macOS animates window close/hide (~250ms) and only removes the window from
the screen once the process services its main run loop; a Jupyter kernel
never does after the cell ends, so the dead window lingers until the kernel
is killed (pygame-ce #2256, libsdl SDL#10081).

Use quit_pygame() instead of pygame.quit(). On non-mac platforms it is just
pygame.quit().
"""
import sys

import pygame


def quit_pygame():
    if sys.platform != "darwin":
        pygame.quit()
        return

    import ctypes
    import ctypes.util

    objc = ctypes.CDLL(ctypes.util.find_library("objc"))
    ctypes.CDLL(ctypes.util.find_library("AppKit"))  # ensure AppKit is loaded
    cf = ctypes.CDLL(ctypes.util.find_library("CoreFoundation"))

    objc.objc_getClass.restype = ctypes.c_void_p
    objc.sel_registerName.restype = ctypes.c_void_p
    send = objc.objc_msgSend

    def msg(receiver, selname, *args, restype=ctypes.c_void_p, argtypes=()):
        send.restype = restype
        send.argtypes = [ctypes.c_void_p, ctypes.c_void_p] + list(argtypes)
        sel = objc.sel_registerName(selname.encode())
        return send(ctypes.c_void_p(receiver), ctypes.c_void_p(sel), *args)

    # Disable the close/hide animation on every window of this process
    # (NSWindowAnimationBehaviorNone == 2). Must happen before the windows
    # are ordered out, i.e. before pygame.quit().
    nsapp = msg(objc.objc_getClass(b"NSApplication"), "sharedApplication")
    windows = msg(nsapp, "windows")
    for i in range(msg(windows, "count", restype=ctypes.c_ulong)):
        w = msg(windows, "objectAtIndex:", ctypes.c_ulong(i),
                argtypes=[ctypes.c_ulong])
        msg(w, "setAnimationBehavior:", ctypes.c_long(2),
            restype=None, argtypes=[ctypes.c_long])

    pygame.quit()

    # One short main-runloop pass so AppKit commits the deferred order-out.
    cf.CFRunLoopRunInMode.restype = ctypes.c_int32
    cf.CFRunLoopRunInMode.argtypes = [ctypes.c_void_p, ctypes.c_double,
                                      ctypes.c_bool]
    mode = ctypes.c_void_p.in_dll(cf, "kCFRunLoopDefaultMode")
    cf.CFRunLoopRunInMode(mode, 0.05, False)
```

**Comment by traversaro** [comment]
> @traversaro I got bored today so got fable to identify the root cause libsdl-org/SDL#15992

Cool, thanks!
